functional_programming.ipynb

```diff
@@ -0,0 +1,2835 @@
+    }
+   },
+   "outputs": [],
+   "source": [
+    "a, *b = [1, 2, 3]"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {
+    "slideshow": {
+     "slide_type": "notes"
+    }
+   },
+   "source": [
+    "If we use `*` before an argument in a function call, this will unpack the argument and pass each component to the function separately, as you can see in this example:"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": 52,
+   "metadata": {
+    "slideshow": {
+     "slide_type": "notes"
+    }
+   },
+   "outputs": [
+    {
+     "name": "stdout",
+     "output_type": "stream",
+     "text": [
+      "a is 1, b is 2, c is 3 \n",
+      "a is 1, b is 2, c is 3 \n",
+      "a is [1, 2, 3], b is [4, 5, 6], c is [5, 6, 7] \n"
+     ]
+    }
+   ],
+   "source": [
+    "def print_variables(a: any, b: any, c: any):\n",
+    "    print(f\"a is {a}, b is {b}, c is {c} \")\n",
+    "\n",
+    "print_variables(*[1, 2, 3])\n",
+    "print_variables(1, 2, 3)\n",
+    "print_variables(*[[1, 2, 3], [4, 5, 6], [5, 6, 7]])"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {
+    "slideshow": {
+     "slide_type": "notes"
+    }
+   },
+   "source": [
+    "Be careful, this only works if the list/iterable you unpack is as long as the arguments you want to pass to your function."
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {
+    "slideshow": {
+     "slide_type": "slide"
+    }
+   },
+   "source": [
+    "\n",
+    "\n",
+    "## Exercises\n",
+    "\n",
+    "\n",
+    "### Exercise 1: Pure or impure functions? 🌶️\n",
+    "\n",
+    "For each of the functions below, determine whether they are pure or impure"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": 53,
+   "metadata": {
+    "slideshow": {
+     "slide_type": "subslide"
+    }
+   },
+   "outputs": [],
+   "source": [
+    "def fun1(a: \"list[str]\") -> None:\n",
+    "    a.append(\"b\")\n",
+    "\n",
+    "\n",
+    "def fun2(a: int) -> int:\n",
+    "    return a + 2\n",
+    "\n",
+    "\n",
+    "def fun3(a: \"dict[str, str]\"):\n",
+    "    a[\"test\"] = \"dest\"\n",
+    "\n"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {
+    "slideshow": {
+     "slide_type": "slide"
+    }
+   },
+   "source": [
+    "### Exercise 2: Keeping only multiples of n 🌶️\n",
+    "\n",
+    "Given a list `L` of integers, write a function that only keeps the numbers that are multiples of a given constant `k`.\n",
+    "\n",
+    "- Example 1: given `nums = [1, 2, 3, 4, 5]`, and `k = 2`, the result must be `[2, 4]`\n",
+    "\n",
+    "- Example 2: given `nums = [1, 2, 3, 4, 5, 6, 7, 8, 9, 10] ` and `k = 5`, the result is `[5, 10]`\n",
+    "\n",
+    "\n",
+    "\n",
+    "<div class=\"alert alert-block alert-info\">\n",
+    "    Hints:\n",
+    "    <ul>\n",
+    "        <li>\n",
+    "            The operator `%` (modulo) gives the remainder of division. `a % k` is 0 when a is a multiple of k. \n",
```

Comment:
Here (and further on) you need to use `<code>` tags, otherwise simple Markdown doesn't work inside these HTML blocks

Reply:
Done in 949b90ffc114ad881542e9d47e8eea8c001ab9dd